ethexe/service/src/fast_sync.rs

```diff
@@ -98,81 +107,155 @@ impl EventData {
-            .context("program states of latest computed block not found")?
-            .into_iter()
-            .map(|(program_id, state)| (program_id, state.hash))
-            .collect();
-
-        for (program_id, state) in computed_program_states {
-            program_states.entry(program_id).or_insert(state);
-        }
+        let latest_committed_block =
+            Self::get_block_data(observer, db, latest_committed_block).await?;
 
-        #[cfg(debug_assertions)]
         if let Some(previous_committed_block) = previous_committed_block {
-            let latest_block_header = db
-                .block_header(latest_committed_block)
-                .expect("observer must fulfill database");
-            let previous_block_header = db
-                .block_header(previous_committed_block)
-                .expect("observer must fulfill database");
-            assert!(previous_block_header.height < latest_block_header.height);
+            let previous_committed_block =
+                Self::get_block_data(observer, db, previous_committed_block).await?;
+            debug_assert!(
+                previous_committed_block.header.height < latest_committed_block.header.height
+            );
         }
 
         Ok(Some(Self {
-            program_states,
-            program_code_ids,
-            needs_instrumentation_codes,
             latest_committed_batch,
             latest_committed_block,
             previous_committed_block,
         }))
     }
 }
 
+async fn net_fetch(
+    network: &mut NetworkService,
+    request: db_sync::Request,
+) -> Result<db_sync::Response> {
+    let request_id = network.db_sync().request(request);
+    loop {
+        let event = network
+            .next()
+            .await
+            .expect("network service stream is infinite");
+
+        if let NetworkEvent::DbResponse {
+            request_id: rid,
+            result,
+        } = event
+        {
+            debug_assert_eq!(rid, request_id, "unknown request id");
+            match result {
+                Ok(response) => break Ok(response),
+                Err((request, err)) => {
+                    log::warn!("Request {:?} failed: {err}. Retrying...", request.id());
+                    network.db_sync().retry(request);
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+async fn collect_program_code_ids(
+    observer: &mut ObserverService,
+    network: &mut NetworkService,
+    latest_committed_block: H256,
+) -> Result<BTreeMap<ActorId, CodeId>> {
+    let router_query = observer.router_query();
+    let programs_count = router_query
+        .programs_count_at(latest_committed_block)
+        .await?;
+
+    let response = net_fetch(
+        network,
+        db_sync::Request::program_ids(latest_committed_block, programs_count),
+    )
+    .await?;
+
+    let program_code_ids = response.unwrap_program_ids();
+    Ok(program_code_ids)
+}
+
+async fn collect_code_ids(
```

Comment:
Some comments on each fn would be good. As far as I see this returns not just codes, but codes to be instrumented?